server/src/main/java/org/apache/druid/segment/loading/SegmentLocalCacheManager.java

```diff
@@ -225,22 +228,49 @@ public List<DataSegment> getCachedSegments() throws IOException
       }
     }
 
-    final List<DataSegment> cachedSegments = new ArrayList<>();
+    final ConcurrentLinkedQueue<DataSegment> cachedSegments = new ConcurrentLinkedQueue<>();
     final File[] segmentsToLoad = retrieveSegmentMetadataFiles();
+    CountDownLatch latch = new CountDownLatch(segmentsToLoad.length);
+
+    boolean createdNewExecutorServiceToLoadSegmentCache = loadOnBootstrapExec == null;
+    ExecutorService executorService = createdNewExecutorServiceToLoadSegmentCache
+                                      ? MoreExecutors.newDirectExecutorService()
+                                      : loadOnBootstrapExec;
 
     AtomicInteger ignoredFilesCounter = new AtomicInteger(0);
 
-    for (int i = 0; i < segmentsToLoad.length; i++) {
-      final File file = segmentsToLoad[i];
-      log.info("Loading segment cache file [%d/%d][%s].", i + 1, segmentsToLoad.length, file);
-      try {
-        addFilesToCachedSegments(file, ignoredFilesCounter, cachedSegments);
-      }
-      catch (Exception e) {
-        log.makeAlert(e, "Failed to load segment from segment cache file.")
-           .addData("file", file)
-           .emit();
-      }
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    log.info("Retrieving [%d] cached segment metadata files to cache.", segmentsToLoad.length);
+
+    for (File file : segmentsToLoad) {
+      executorService.submit(() -> {
+        try {
+          addFilesToCachedSegments(file, ignoredFilesCounter, cachedSegments);
+        }
+        catch (Exception e) {
+          log.makeAlert(e, "Failed to load segment from segment cache file.")
+             .addData("file", file)
+             .emit();
+        }
+        finally {
+          latch.countDown();
+        }
+      });
+    }
+
+    try {
+      latch.await();
+    }
+    catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      log.error(e, "Interrupted when trying to retrieve cached segment metadata files");
```

Comment:
Since it was interrupted, I think we can skip logging the stack trace.
```suggestion
      log.noStackTrace().error(e, "Interrupted when trying to retrieve cached segment metadata files");
```